libcst/codemod/visitors/_apply_type_annotations.py

```diff
@@ -0,0 +1,563 @@
+        self.qualifier: List[str] = []
+        # Store the annotations.
+        self.function_annotations: Dict[str, FunctionAnnotation] = {}
+        self.attribute_annotations: Dict[str, cst.Annotation] = {}
+        self.imports: Dict[str, ImportStatement] = {}
+        self.existing_imports: Set[str] = existing_imports
+        self.class_definitions: Dict[str, cst.ClassDef] = {}
+
+    def visit_ClassDef(self, node: cst.ClassDef) -> None:
+        self.qualifier.append(node.name.value)
+        self.class_definitions[node.name.value] = node
+
+    def leave_ClassDef(self, original_node: cst.ClassDef) -> None:
+        self.qualifier.pop()
+
+    def visit_FunctionDef(self, node: cst.FunctionDef) -> bool:
+        self.qualifier.append(node.name.value)
+        returns = node.returns
+        if returns is not None:
+            return_annotation = self._create_import_from_annotation(returns)
+            parameter_annotations = self._import_parameter_annotations(node.params)
+            self.function_annotations[".".join(self.qualifier)] = FunctionAnnotation(
+                parameters=parameter_annotations, returns=return_annotation
+            )
+        # pyi files don't support inner functions, return False to stop the traversal.
+        return False
+
+    def leave_FunctionDef(self, original_node: cst.FunctionDef) -> None:
+        self.qualifier.pop()
+
+    def visit_AnnAssign(self, node: cst.AnnAssign) -> bool:
+        name = _get_name_or_attribute_as_string(node.target)
+        if name is not None:
+            self.qualifier.append(name)
+        annotation_value = self._create_import_from_annotation(node.annotation)
+        self.attribute_annotations[".".join(self.qualifier)] = annotation_value
+        return True
+
+    def leave_AnnAssign(self, original_node: cst.AnnAssign) -> None:
+        self.qualifier.pop()
+
+    def visit_ImportFrom(self, node: cst.ImportFrom) -> None:
+        module = node.module
+        names = node.names
+
+        # module is None for relative imports like `from .. import foo`.
+        # We ignore these for now.
+        if module is None or isinstance(names, cst.ImportStar):
+            return
+        name = _get_name_or_attribute_as_string(module)
+        if name is not None:
+            self._add_to_imports(names, cst.Name(name), name)
+
+    def _add_annotation_to_imports(
+        self, annotation: cst.Attribute
+    ) -> Union[cst.Name, cst.Attribute]:
+        key = _get_name_or_attribute_as_string(annotation.value)
+        if key is not None:
+            # Don't attempt to re-import existing imports.
+            if key in self.existing_imports:
+                return annotation
+            value = annotation.value
+            if isinstance(value, (cst.Attribute, cst.Name)):
+                self._add_to_imports(
+                    [cst.ImportAlias(name=annotation.attr)], value, key
+                )
+        return annotation.attr
```

Comment:
I'm really sorry that just realized a lot of code here are for handling adding missing import.
Good news is, we already have that implemented as a helper `AddImportsVisitor`.
https://libcst.readthedocs.io/en/latest/codemods.html#libcst.codemod.visitors.AddImportsVisitor

To add `from typing import Optional`,
you just need to call: `AddImportsVisitor.add_needed_import(self.context, "typing", "Optional")`.

To add `import typing`,
you just need to call: `AddImportsVisitor.add_needed_import(self.context, "typing")`.

`AddImportsVisitor` will do the work to make sure there is no duplicated imports for you. So a lot of code in `TypeCollector` and `ApplyTypeAnnotationsVisitor` can be saved. You also don't need to use `GatherImportsVisitor` to collect imports. You can safely add all import needed by type annotation.
I should have point this out sooner.